src/utils.jl

```diff
@@ -1,14 +1,98 @@
 ## macro for argument checking
 
-macro check_args(D, cond)
-    quote
-        if !($(esc(cond)))
-            throw(ArgumentError(string(
-                $(string(D)), ": the condition ", $(string(cond)), " is not satisfied.")))
+macro strip_linenos(expr)
+    return esc(Base.remove_linenums!(expr))
+end
+
+"""
+    @check_args(
+        D,
+        @setup(statements...),
+        (arg₁, cond₁, message₁),
+        (cond₂, message₂),
+        ...,
+    )
+
+A convenience macro that generates AD-compatible checks of arguments for a distribution of
+type `D`.
+
+More concretely, it generates the following Julia code:
+```julia
+Distributions.check_args(check_args) do
+    \$(statements...)
+    cond₁ || throw(DomainError(arg₁, \$(string(D, ": ", message₁))))
+    cond₂ || throw(ArgumentError(\$(string(D, ": ", message₂))))
+    ...
+end
+```
+
+The `@setup` argument can be elided if no setup code is needed. Moreover, error messages
+can be omitted. In this case the message `"the condition \$(cond) is not satisfied."` is
+used.
+"""
+macro check_args(D, setup_or_check, checks...)
+    # Extract setup statements
+    if Meta.isexpr(setup_or_check, :macrocall) && setup_or_check.args[1] == Symbol("@setup")
+        setup_stmts = Any[esc(ex) for ex in setup_or_check.args[3:end]]
+    else
+        setup_stmts = []
+        checks = (setup_or_check, checks...)
+    end
+
+    # Generate expressions for each condition
+    conds_exprs = map(checks) do check
+        if Meta.isexpr(check, :tuple, 3)
+            # argument, condition, and message specified
+            arg = check.args[1]
+            cond = check.args[2]
+            message = string(D, ": ", check.args[3])
+            return :(($(esc(cond))) || throw(DomainError($(esc(arg)), $message)))
+        elseif Meta.isexpr(check, :tuple, 2)
+            cond_or_message = check.args[2]
+            if cond_or_message isa String
+                # only condition and message specified
+                cond = check.args[1]
+                message = string(D, ": ", cond_or_message)
+                return :(($(esc(cond))) || throw(ArgumentError($message)))
+            else
+                # only argument and condition specified
+                arg = check.args[1]
+                cond = cond_or_message
+                message = string(D, ": the condition ", cond, " is not satisfied.")
+                return :(($(esc(cond))) || throw(DomainError($(esc(arg)), $message)))
+            end
+        else
+            # only condition specified
+            cond = check
+            message = string(D, ": the condition ", cond, " is not satisfied.")
+            return :(($(esc(cond))) || throw(ArgumentError($message)))
+        end
+    end
```

Comment:
The idea here is to throw a `DomainError` if possible since it yields more descriptive error messages that also display the problematic value without having to construct strings dynamically which breaks GPU support (https://github.com/JuliaStats/Distributions.jl/pull/1491).